File: client/config/config_loader.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
LJWX系统配置加载器
统一配置管理，支持YAML配置文件和环境变量覆盖
"""

import os
import yaml
from typing import Dict, Any, Optional
from pathlib import Path


class LJWXConfigLoader:
    """LJWX配置加载器"""
    
    def __init__(self, config_path: Optional[str] = None):
        """
        初始化配置加载器
        
        Args:
            config_path: 配置文件路径，默认为当前目录下的ljwx-config.yaml
        """
        if config_path is None:
            # 尝试多个可能的配置文件位置
            possible_paths = [
                "/app/config/ljwx-config.yaml",  # 容器内路径
                "/client/config/ljwx-config.yaml",  # 挂载路径
                os.path.join(os.path.dirname(__file__), "ljwx-config.yaml"),  # 同级目录
                "client/config/ljwx-config.yaml",  # 相对路径
                "config/ljwx-config.yaml"  # 简化路径
            ]
            
            config_path = None
            for path in possible_paths:
                if os.path.exists(path):
                    config_path = path
                    break
            
            if config_path is None:
                raise FileNotFoundError("配置文件未找到，请确保ljwx-config.yaml存在")
        
        self.config_path = config_path
        self.config = self._load_config()
    
    def _load_config(self) -> Dict[str, Any]:
        """加载配置文件"""
        try:
            with open(self.config_path, 'r', encoding='utf-8') as f:
                config = yaml.safe_load(f)
            
            # 环境变量覆盖
            config = self._apply_env_overrides(config)
            
            return config
        except Exception as e:
            raise Exception(f"加载配置文件失败: {e}")
    
    def _apply_env_overrides(self, config: Dict[str, Any]) -> Dict[str, Any]:
        """应用环境变量覆盖"""
        # 数据库配置覆盖
        if 'MYSQL_HOST' in os.environ:
            config['database']['mysql']['host'] = os.environ['MYSQL_HOST']
        if 'MYSQL_PORT' in os.environ:
            config['database']['mysql']['port'] = int(os.environ['MYSQL_PORT'])
        if 'MYSQL_DATABASE' in os.environ:
            config['database']['mysql']['database'] = os.environ['MYSQL_DATABASE']
        if 'MYSQL_USERNAME' in os.environ:
            config['database']['mysql']['username'] = os.environ['MYSQL_USERNAME']
        if 'MYSQL_PASSWORD' in os.environ:
            config['database']['mysql']['password'] = os.environ['MYSQL_PASSWORD']
        
        if 'REDIS_HOST' in os.environ:
            config['database']['redis']['host'] = os.environ['REDIS_HOST']
        if 'REDIS_PORT' in os.environ:
            config['database']['redis']['port'] = int(os.environ['REDIS_PORT'])
        if 'REDIS_PASSWORD' in os.environ:
            config['database']['redis']['password'] = os.environ['REDIS_PASSWORD']
        
        # 服务端口覆盖
        if 'SERVER_PORT' in os.environ:
            service_name = os.environ.get('SERVICE_NAME', 'ljwx-boot')
            if service_name in config['services']:
                config['services'][service_name]['port'] = int(os.environ['SERVER_PORT'])
        
        return config
    
    def get(self, key: str, default: Any = None) -> Any:
        """
        获取配置值，支持点号分隔的嵌套键
        
        Args:
            key: 配置键，如 'database.mysql.host'
            default: 默认值
            
        Returns:
            配置值
        """
        keys = key.split('.')
        value = self.config
        
        for k in keys:
            if isinstance(value, dict) and k in value:
                value = value[k]
            else:
                return default
        
        return value
    
    def get_database_config(self, db_type: str = 'mysql') -> Dict[str, Any]:
        """获取数据库配置"""
        return self.get(f'database.{db_type}', {})
    
    def get_service_config(self, service_name: str) -> Dict[str, Any]:
        """获取服务配置"""
        return self.get(f'services.{service_name}', {})
    
    def get_image_config(self) -> Dict[str, str]:
        """获取镜像配置"""
        return self.get('images', {})
    
    def get_database_url(self, db_type: str = 'mysql') -> str:
        """获取数据库连接URL"""
        db_config = self.get_database_config(db_type)
        
        if db_type == 'mysql':
            return f"mysql+pymysql://{db_config['username']}:{db_config['password']}@{db_config['host']}:{db_config['port']}/{db_config['database']}?charset={db_config.get('charset', 'utf8mb4')}"
        elif db_type == 'redis':
            password_part = f":{db_config['password']}@" if db_config.get('password') else ""
            return f"redis://{password_part}{db_config['host']}:{db_config['port']}/{db_config.get('db', 0)}"
        
        return ""
    
    def reload(self):
        """重新加载配置"""
        self.config = self._load_config()


# 全局配置实例
_config_loader = None

def get_config() -> LJWXConfigLoader:
    """获取全局配置实例"""
    global _config_loader
    if _config_loader is None:
        _config_loader = LJWXConfigLoader()
    return _config_loader

def reload_config():
    """重新加载全局配置"""
    global _config_loader
    if _config_loader is not None:
        _config_loader.reload()

# 便捷函数
def get_database_config(db_type: str = 'mysql') -> Dict[str, Any]:
    """获取数据库配置"""
    return get_config().get_database_config(db_type)

def get_service_config(service_name: str) -> Dict[str, Any]:
    """获取服务配置"""
    return get_config().get_service_config(service_name)

def get_database_url(db_type: str = 'mysql') -> str:
    """获取数据库连接URL"""
    return get_config().get_database_url(db_type)


if __name__ == "__main__":
    # 测试配置加载
    config = get_config()
    
    print("=== LJWX配置测试 ===")
    print(f"系统名称: {config.get('system.name')}")
    print(f"系统版本: {config.get('system.version')}")
    print()
    
    print("=== 数据库配置 ===")
    mysql_config = get_database_config('mysql')
    print(f"MySQL: {mysql_config}")
    print(f"MySQL URL: {get_database_url('mysql')}")
    print()
    
    redis_config = get_database_config('redis')  
    print(f"Redis: {redis_config}")
    print(f"Redis URL: {get_database_url('redis')}")
    print()
    
    print("=== 服务配置 ===")
    for service in ['ljwx-admin', 'ljwx-bigscreen', 'ljwx-boot']:
        service_config = get_service_config(service)
        print(f"{service}: {service_config}")
    print()
    
    print("=== 镜像配置 ===")
    image_config = config.get_image_config()
    print(f"镜像仓库: {image_config.get('registry')}")
    print(f"Admin镜像: {image_config.get('ljwx-admin')}")
    print(f"Bigscreen镜像: {image_config.get('ljwx-bigscreen')}")
    print(f"Boot镜像: {image_config.get('ljwx-boot')}")

File: client/config/config-loader.js
/**
 * LJWX系统配置加载器 - JavaScript版本
 * 统一配置管理，支持YAML配置文件和环境变量覆盖
 * 
 * @author LJWX Team
 * @version 2.0.1
 */

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');

class LJWXConfigLoader {
    constructor(configPath = null) {
        this.configPath = configPath;
        this.config = null;
        this.loadConfig();
    }

    /**
     * 加载配置文件
     */
    loadConfig() {
        const possiblePaths = [
            '/app/config/ljwx-config.yaml',  // 容器内路径
            '/client/config/ljwx-config.yaml',  // 挂载路径
            path.join(__dirname, 'ljwx-config.yaml'),  // 同级目录
            'client/config/ljwx-config.yaml',  // 相对路径
            'config/ljwx-config.yaml',  // 简化路径
            this.configPath  // 指定路径
        ].filter(p => p);

        for (const configPath of possiblePaths) {
            if (fs.existsSync(configPath)) {
                try {
                    const fileContents = fs.readFileSync(configPath, 'utf8');
                    this.config = yaml.load(fileContents);
                    
                    // 应用环境变量覆盖
                    this.applyEnvOverrides();
                    return;
                } catch (error) {
                    // 继续尝试下一个路径
                    continue;
                }
            }
        }
        
        throw new Error('配置文件未找到，请确保ljwx-config.yaml存在');
    }

    /**
     * 应用环境变量覆盖
     */
    applyEnvOverrides() {
        // 数据库配置覆盖
        if (process.env.MYSQL_HOST) {
            this.config.database.mysql.host = process.env.MYSQL_HOST;
        }
        if (process.env.MYSQL_PORT) {
            this.config.database.mysql.port = parseInt(process.env.MYSQL_PORT);
        }
        if (process.env.MYSQL_DATABASE) {
            this.config.database.mysql.database = process.env.MYSQL_DATABASE;
        }
        if (process.env.MYSQL_USERNAME) {
            this.config.database.mysql.username = process.env.MYSQL_USERNAME;
        }
        if (process.env.MYSQL_PASSWORD) {
            this.config.database.mysql.password = process.env.MYSQL_PASSWORD;
        }

        if (process.env.REDIS_HOST) {
            this.config.database.redis.host = process.env.REDIS_HOST;
        }
        if (process.env.REDIS_PORT) {
            this.config.database.redis.port = parseInt(process.env.REDIS_PORT);
        }
        if (process.env.REDIS_PASSWORD) {
            this.config.database.redis.password = process.env.REDIS_PASSWORD;
        }

        // 服务端口覆盖
        if (process.env.SERVER_PORT && process.env.SERVICE_NAME) {
            const serviceName = process.env.SERVICE_NAME;
            if (this.config.services[serviceName]) {
                this.config.services[serviceName].port = parseInt(process.env.SERVER_PORT);
            }
        }
    }

    /**
     * 获取配置值，支持点号分隔的嵌套键
     * 
     * @param {string} key - 配置键，如 'database.mysql.host'
     * @param {any} defaultValue - 默认值
     * @returns {any} 配置值
     */
    get(key, defaultValue = null) {
        const keys = key.split('.');
        let value = this.config;

        for (const k of keys) {
            if (value && typeof value === 'object' && k in value) {
                value = value[k];
            } else {
                return defaultValue;
            }
        }

        return value;
    }

    /**
     * 获取数据库配置
     * 
     * @param {string} dbType - 数据库类型，如 'mysql', 'redis'
     * @returns {object} 数据库配置对象
     */
    getDatabaseConfig(dbType = 'mysql') {
        return this.get(`database.${dbType}`, {});
    }

    /**
     * 获取服务配置
     * 
     * @param {string} serviceName - 服务名称
     * @returns {object} 服务配置对象
     */
    getServiceConfig(serviceName) {
        return this.get(`services.${serviceName}`, {});
    }

    /**
     * 获取镜像配置
     * 
     * @returns {object} 镜像配置对象
     */
    getImageConfig() {
        return this.get('images', {});
    }

    /**
     * 获取数据库连接URL
     * 
     * @param {string} dbType - 数据库类型
     * @returns {string} 数据库连接URL
     */
    getDatabaseUrl(dbType = 'mysql') {
        const dbConfig = this.getDatabaseConfig(dbType);

        if (dbType === 'mysql') {
            const { host, port, database, username, password, charset = 'utf8mb4' } = dbConfig;
            return `mysql://${username}:${password}@${host}:${port}/${database}?charset=${charset}`;
        } else if (dbType === 'redis') {
            const { host, port, password = '', db = 0 } = dbConfig;
            const passwordPart = password ? `:${password}@` : '';
            return `redis://${passwordPart}${host}:${port}/${db}`;
        }

        return '';
    }

    /**
     * 重新加载配置
     */
    reload() {
        this.loadConfig();
    }
}

// 全局配置实例
let _configLoader = null;

/**
 * 获取全局配置实例
 * 
 * @returns {LJWXConfigLoader} 配置加载器实例
 */
function getConfig() {
    if (_configLoader === null) {
        _configLoader = new LJWXConfigLoader();
    }
    return _configLoader;
}

/**
 * 重新加载全局配置
 */
function reloadConfig() {
    if (_configLoader !== null) {
        _configLoader.reload();
    }
}

// 便捷函数
/**
 * 获取数据库配置
 */
function getDatabaseConfig(dbType = 'mysql') {
    return getConfig().getDatabaseConfig(dbType);
}

/**
 * 获取服务配置
 */
function getServiceConfig(serviceName) {
    return getConfig().getServiceConfig(serviceName);
}

/**
 * 获取数据库连接URL
 */
function getDatabaseUrl(dbType = 'mysql') {
    return getConfig().getDatabaseUrl(dbType);
}

module.exports = {
    LJWXConfigLoader,
    getConfig,
    reloadConfig,
    getDatabaseConfig,
    getServiceConfig,
    getDatabaseUrl
};

// 如果是直接运行此文件，进行测试
if (require.main === module) {
    try {
        const config = getConfig();
        
        console.log('=== LJWX配置测试 ===');
        console.log(`系统名称: ${config.get('system.name')}`);
        console.log(`系统版本: ${config.get('system.version')}`);
        console.log('');
        
        console.log('=== 数据库配置 ===');
        const mysqlConfig = getDatabaseConfig('mysql');
        console.log(`MySQL:`, mysqlConfig);
        console.log(`MySQL URL: ${getDatabaseUrl('mysql')}`);
        console.log('');
        
        const redisConfig = getDatabaseConfig('redis');
        console.log(`Redis:`, redisConfig);
        console.log(`Redis URL: ${getDatabaseUrl('redis')}`);
        console.log('');
        
        console.log('=== 服务配置 ===');
        ['ljwx-admin', 'ljwx-bigscreen', 'ljwx-boot'].forEach(service => {
            const serviceConfig = getServiceConfig(service);
            console.log(`${service}:`, serviceConfig);
        });
        console.log('');
        
        console.log('=== 镜像配置 ===');
        const imageConfig = config.getImageConfig();
        console.log(`镜像仓库: ${imageConfig.registry}`);
        console.log(`Admin镜像: ${imageConfig['ljwx-admin']}`);
        console.log(`Bigscreen镜像: ${imageConfig['ljwx-bigscreen']}`);
        console.log(`Boot镜像: ${imageConfig['ljwx-boot']}`);
        
    } catch (error) {
        console.error('配置加载测试失败:', error.message);
    }
}